[stm32] PM restore console after sleep mode

FRASTM opened the pull request:
With this change, the clock configuration is restored when leaving the low power mode. 
Then the console of the nucleo_wb55 can operate after sleeping.

The UART executes the `uart_stm32_set_power_state` to have a clean configuration before entering the low power mode.
On wakeup, the clocks are re-configured thanks to the `stm32_clock_control_init` function called directly in the 
`pm_power_state_exit_post_ops` of each soc.
This method is more efficient than introducing a _clock_stm32_pm_control_ function in the clock driver. 

Fix #31284
Requires #32042

Signed-off-by: Francois Ramu francois.ramu@st.com

FRASTM committed:
drivers: serial: stm32 restore uart after lowpower

This adds a function to control the uart device during lowpower modes

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: clock: stm32 clock control with low power modes

This will export the stm32_clock_control_init function
to restore the clocks after the low power modes.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
soc: arm: stm32wb low power mode restoring clocks

The clock configuration is restored by the stm32_clock_control_init
function of the clock_stm32_ll_common driver
when exiting the low power mode.
Keeping the stm32wbxx_ll_utils.h in case the PLL is used for wb.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
*** Booting Zephyr OS build v2.5.0-rc1-116-g9d9c34e27c13  ***                                                                           
Running test suite power_management_test                                                                                                
===================================================================                                                                     
START - test_power_idle                                                                                                                 
give way to idle thread                                                                                                                 
 PASS - test_power_idle                                                                                                                 
===================================================================                                                                     
START - test_power_state_trans                                                                                                          
 PASS - test_power_state_trans                                                                                                          
===================================================================                                                                     
START - test_power_state_notification                                                                                                   
 PASS - test_power_state_notification                                                                                                   
===================================================================                                                                     
Test suite power_management_test succeeded                                                                                              
===================================================================                                                                     
PROJECT EXECUTION SUCCESSFUL

tagunil commented on drivers/serial/uart_stm32.c:
```diff
@@ -52,6 +55,10 @@ LOG_MODULE_REGISTER(uart_stm32);
 #define TIMEOUT 1000
 
+#ifdef CONFIG_PM_DEVICE
+char tmp;
```
What's the point of this variable? Is there some weird reading optimization problem?

gmarull replied:
Its scope could likely be at `while (LL_USART_IsActiveFlag_RXNE(UartInstance)) {` loop

FRASTM replied:
var is used to flush the rx fifo, now moved in the function itself

gmarull commented on drivers/clock_control/clock_stm32_ll_common.h:
```diff
@@ -30,10 +30,15 @@
 void LL_RCC_MSI_Disable(void);
 
+
+int stm32_clock_control_init(const struct device *dev);
```
If made public it's worth adding some docstrings

FRASTM replied:
ok, adding some lines

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -35,6 +35,9 @@
 #include <stm32_ll_usart.h>
 #include <stm32_ll_lpuart.h>
+#ifdef CONFIG_PM_DEVICE
+#include <stm32_ll_rcc.h>
```
Leftover ?

FRASTM commented:
Remove the commit  to` soc: arm: stm32wb low power mode for CPU2 too` . The CPU2 lowpower mode is not set here (LL_C2_PWR_SetPowerMode(LL_PWR_MODE_STOPx);
and changes after review

gmarull commented on drivers/serial/uart_stm32.c:
```diff
@@ -1389,6 +1392,88 @@ static int uart_stm32_init(const struct device *dev)
+			data->pm_state = new_state;
+		}
+	} else {
+		__ASSERT_NO_MSG(ctrl_command == DEVICE_PM_GET_POWER_STATE);
```
It may be part of an `} else if () {`, then just return for the rest. If any `ctrl_command` gets added, this code could still be valid but would fail due to this assertion.

FRASTM replied:
simplification in  testing only new_state != DEVICE_PM_ACTIVE_STATE

FRASTM committed:
soc: arm: stm32 devices supporting low power modes

stm32wb and stm32l4 are devices which supports the low power
from the Kernel PM, so is the PM_DEVICE defined.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -1389,6 +1392,88 @@ static int uart_stm32_init(const struct device *dev)
+		/* nothing special to do on UartInstance */
+		data->pm_state = new_state;
+	} else {
+		__ASSERT_NO_MSG(new_state == DEVICE_PM_LOW_POWER_STATE ||
```
not really, removed

FRASTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -1389,6 +1392,88 @@ static int uart_stm32_init(const struct device *dev)
+static int uart_stm32_set_power_state(const struct device *dev,
+					      uint32_t new_state)
+{
+	char tmp;
```
yes, now changed

FRASTM replied:
uart_stm32_set_power_state simplified
adding the power.c for stm32l4 : restoring the clock with `stm32_clock_control_init`

FRASTM committed:
soc: arm: stm32l4 low power mode restoring clocks

The clock configuration is restored by the stm32_clock_control_init
function of the clock_stm32_ll_common driver
when exiting the low power mode

Signed-off-by: Francois Ramu <francois.ramu@st.com>


Signed-off-by: Francois Ramu <francois.ramu@st.com>

dcpleung commented on drivers/serial/uart_stm32.c:
```diff
@@ -1387,6 +1390,65 @@ static int uart_stm32_init(const struct device *dev)
 }
 
+#ifdef CONFIG_PM_DEVICE
+static int uart_stm32_set_power_state(const struct device *dev,
```
This does not _exactly_ set the power state, which is confusing to future readers. I suggest moving the `data->pm_state = new_state` from below to here, or rename this function to something like `uart_stm32_prep_power_transition()`.

FRASTM replied:
However it seems to be more or less generic that the `uart_xx_pm_control` is calling a `uart_xx_set_power_state` function.
The test in the uart_stm32_set_power_state function is just checking that the UART is requested to enter a _non-DEVICE_PM_ACTIVE_STATE_

dcpleung replied:
Only `uart_nrfx_uart.c` is setting the field in device struct outside of `uart_*_set_power_state()`. The other two drivers which have PM enabled, `uart_cc13xx_cc26xx.c` and `uart_nrfx_uarte.c`, are doing this inside the setter function. I still think it is better to actually set the state inside the setter function to be less confusing.

abelNoelleC replied:
I'm personally still pro at least renaming the function if you want it to keep it separate.

abel-NoelleC commented:
Thank you for this PR @FRASTM ! It would help me a lot if it was merged since it's required for STM32L1x power management too (when using the stop mode).

erwango commented:
> Thank you for this PR @FRASTM ! It would help me a lot if this was merged since it's required for STM32L1x power management too (when using the stop mode).

@abel-NoelleC The best way to help in this direction is review, and approve when you're satisfied (in the scope on the proposal)

abel-NoelleC commented on drivers/serial/uart_stm32.c:
```diff
@@ -1387,6 +1390,65 @@ static int uart_stm32_init(const struct device *dev)
 #endif
 }
 
+#ifdef CONFIG_PM_DEVICE
```
`uart_stm32_pm_control()` is now also part of this #ifdef. Could this cause problems when `CONFIG_PM_DEVICE` isn't selected?

FRASTM replied:
In case the PM_DEVICE is not defined, the uart_stm32_pm_control does not exist but  DEVICE_DT_INST_DEFINE get a null value for this entry (no error).

abelNoelleC reviewed (approved):
If @dscpleung 's comment is resolved, and if having `uart_stm32_pm_control()` inside the `CONFIG_PM_DEVICE` bounds doesn't cause problems, this would meet our needs. 

Additionally I'd like to discuss if peripheral (in this case UART) specific GPIO configuration for power management would also fit in these drivers. Not just GPIO port configuration, but also peripheral deinitialization (for example for UART use of `LL_USART_DeInit(USARTx)` function, where peripheral clock is disabled). 
IMO not something that blocks this PR, but something to keep in mind and discuss now that the power management is being added to these peripheral device drivers.

erwango reviewed (changes requested):
Let's add nucleo_wb55rg to `tests/subsys/power/power_mgmt_soc/testcase.yaml`
and we're good to go.

FRASTM committed:
tests: subsys: power: add stm32 targets to the power_mgmt_soc test case

New stm32 nucleo boards with CONFIG_PM_DEVICE can pass this test.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding a commit to include nucleo_wb55rg and nucleo_l476rg boards for the `tests/subsys/power/power_mgmt_soc`

tagunil commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -8,6 +8,10 @@
 #include <init.h>
 
+#include <stm32_ll_rcc.h>
+#include <stm32_ll_utils.h>
```
Why are these two includes needed here, but not in stm32wb/power.c? What's the difference?

FRASTM replied:
The reason is that the clock_stm32_ll_common_.h   includes the LL_RCC_MSI_Disable and LL_UTILS_PLLInitTypeDef
for the stm32l4 soc serie (which is not the case with the wb55)

tagunil replied:
Sorry, but I don't see any includes at all in clock_stm32_ll_common.h.

FRASTM replied:
That's why they are in the power.c.

tagunil replied:
Yes, but then how exactly does stm32wb/power.c work without them? Is it supposed to be this way? Though it's definitely not a blocker.

erwango replied:
It should work the the same as it works on L4 with PLL. But point is main BLE config doesn't use PLL, meaning PLL is not a likely config on WB.

tagunil replied:
@erwango thanks for the clarification!
@FRASTM wouldn't it be more consistent to include <stm32l4xx_ll_utils.h> instead of the generic one?

FRASTM replied:
yes, changing to `stm32l4xx_ll_utils.h`

abelNoelleC replied:
Can someone explain why the non-generic libraries would be more consistent? (Since I also have opened #32322)

tagunil replied:
> Can someone explain why the non-generic libraries would be more consistent? (Since I also have opened #32322)

My point was that all the other similar includes in soc.c files across the series are non-generic, so it would be better to do the same, or to change all of them to generic ones at once.

abelNoelleC replied:
Aaaah alright, thanks!

FRASTM committed:
soc: arm: stm32l4 give a weak low power mode devices for policy

In case the policy is not available or defined by th application
the pm_policy_low_power_devices is still valid

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
soc: arm: stm32wb give a weak low power mode devices for policy

In case the policy is not available or defined by th application
the pm_policy_low_power_devices is still valid

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Add more target for the tests/subsys/power/power_mgmt
Fix compilation build error with tests/kernel/profiling/profiling_api by adding a weak `pm_policy_low_power_devices`, for each stm32 soc (in soc/power.c) in case the policy is not available.

FRASTM commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -8,6 +8,10 @@
 #include <soc.h>
 #include <init.h>
 
+#include <stm32_ll_rcc.h>
```
correct

FRASTM commented:
adding the `include <stm32wbxx_ll_utils.h> for power.c of the stm32l4 and stm32wb, in case the PLL is enable here.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1387,6 +1390,65 @@ static int uart_stm32_init(const struct device *dev)
+		if (new_state != data->pm_state) {
+			uart_stm32_set_power_state(dev, new_state);
+			data->pm_state = new_state;
+		}
```
Nit: Indentation not correct here.

erwango commented on tests/subsys/power/power_mgmt_soc/testcase.yaml:
```diff
@@ -1,4 +1,5 @@
 tests:
   subsys.power.pm_soc:
     platform_allow: cc26x2r1_launchxl cc1352r1_launchxl mec15xxevb_assy6853 mec1501modular_assy6885
+       nucleo_wb55rg nucleo_l4r5zi nucleo_l476rg disco_l475_iot1 b_l4s5i_iot01a
```
Only nucleo_wb55rg nucleo_l476rg supported right now.

FRASTM commented:
rebase

FRASTM commented:
rebase

erwango commented:
@dcpleung can you have a new look ?
